cfs/cmkdir.1
.TH CMKDIR 1 ""
.SH NAME
cmkdir - create encrypted directory for CFS
.SH SYNOPSIS
.B cmkdir
[ \-\fB123bdmosp\fP ]
\fIdirectory\fP
.SH DESCRIPTION
\fBcmkdir\fP creates \fIdirectory\fP and assigns to it cryptographic
keys for use by the Cryptographic File System (CFS).  Operation is
similar to the ordinary \fBmkdir\fP(1) command, with the addition that
the user is prompted for a passphrase which is used to generate the
DES keys used by cfsd(8) to transparently encrypt the files.  The
smartcard version of \fBcmkdir\fP initializes a key smartcard and
requires that a blank smartcard be inserted into the smartcard reader.
.LP
Once created, encrypted directories can be made available for use with
the \fBcattach\fP(1) command.  Users should not ordinarily read and
write directly to directories created with \fBcmkdir\fP, since these
files would not be stored in encrypted form.
.LP
By default, \fBcmkdir\fP creates directories for two-key hybrid mode
triple DES.  The \-\fB1\fP option specifies two-key hybrid mode single
DES; this is faster, albiet at the expense of security.  Three-key
triple DES is specified with \-\fB3\fP; directories created for
three-key triple DES cannot be read by versions of CFS earlier than
1.3.2.  Other cipher algorithms may also be available,
depending on the local configuration.
.LP
Use the -o option to create directories that can be read by versions
of CFS before 1.3; directories created under this option can be read
by
.B cname
and
.B ccat
as well.
.LP
The -p ("puny") option creates directories that use much less memory
when attached under cfsd.  This is useful on machines with very little
(less than, say, 8MBs with a window system and browser also running)
memory.  Files in directories created under -p may reveal slightly
more about their structure than regular CFS files.
.LP
The -- option will read the key from standard input, and will not
attempt to read from /dev/tty or change the terminal modes.  This is
useful for creating directories from other programs or scripts, and
should not ordinarily be used.
.LP
Three new experimental block ciphers are included in the default
distribution.  The \-\fBb\fP oprion specifies Schneier's popular "Blowfish"
algorithm.  It has a 128 bit nominal keyspace and is rather fast on most
computers.  Blowfish is a fairly new algorithm and has not enjoyed nearly
the analytic attention that DES has, so it is not recommended for critical
applications.  The \-\fBm\fP option specifies Blaze and Schneier's
experimental "MacGuffin" cipher.  It has 32 rounds, a 64 bit codebook size
and a 128 bit nominal keyspace.  Use this cipher at your own risk; it is
much weaker than its keyspace suggests, and is included only as
an example.
.LP
Another new cipher, James Massey's SAFER-SK128, is also available in
this release.  Specify SAFER-SK128 with the \-\fBs\fP option.  Again,
this cipher hasn't been around nearly as long as DES, so use it at
your own risk.  SAFER is a little faster than triple DES.
.SH FILES
.TP
\fIdirectory\fP/...
known-plaintext hash of the assigned keys.
.TP
\fIdirectory\fP/..c
identifies the cipher algorithm.
.SH SEE ALSO
cfsd(8), cattach(1)
.SH BUGS
The MacGuffin, Blowfish and SAFER ciphers aren't nearly as
well-studied as DES.  They are included primarly as an example of how
to add ciphers to CFS.  The author's personal files remain protected
with the -2 option.
.LP
Some of the options (-2, -3) have different meanings from previous
versions.
.SH AUTHOR
Matt Blaze; for information on cfs, email to cfs@research.att.com.


rrc2/history
 ############################################################################
#                                                                            #
#                      History of my Implementation of                       #
#                      RRC.2 - Ron Rivest Cipher Nr. 2                       #
#                                                                            #
 ############################################################################

13.02.96:
	Received specification

15.02.96: 
	Linux (ANSI-) C version done;
	includes RRC.2 library, test program, and simple sample crypt
	program

18.02.96:
	First announcement on usenet (rrc2_news.txt)

18.02.96:
	Turbo Pascal version done;
	Includes only Unit RRC2 and test program.

20.02.96:
	Improved C version: now endian neutral (untested)

20.02.96:
	RRC.2 obtainable via ftp

23.02.96:
	second usenet announcement (rrc2_news_2.txt)

# version FINI


cfs/cmkdir.c
/*
 * The author of this software is Matt Blaze.
 *              Copyright (c) 1992, 1993, 1994, 1995, 1997 by AT&T.
 * Permission to use, copy, and modify this software without fee
 * is hereby granted, provided that this entire notice is included in
 * all copies of any software which is or includes a copy or
 * modification of this software and in all copies of the supporting
 * documentation for such software.
 *
 * This software is subject to United States export controls.
 *
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR AT&T MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */

/*
 * client cfs mkdir - 1.4.0
 */
#include <stdio.h>
#include <rpc/rpc.h>
#include <sys/time.h>
#include "nfsproto.h"
#include "admproto.h"
#include "cfs.h"
#include "shs.h"

main(argc,argv)
     int argc;
     char **argv;
{
	char *pw;
	char pword[256];
	char *getpassword();
	int n;
	cfs_admkey k;
	cfskey kt;
	char path[1024];
	char str[8];
	FILE *fp;
	char *flg;
	struct timeval tv;
	u_long r;
	int i;
	int ciph=CFS_THREE_DES;
	int cfmt=1;
	int smsize=LARGESMSIZE;
	unsigned char ekey[128];
	unsigned char ek1[128];
	unsigned int l;
	int keycheck=1;
	
	while (--argc && (**++argv == '-')) {
		for (flg= ++*argv; *flg; ++flg)
			switch (*flg) {
			    case '1':
			    case 'd':
				ciph=CFS_STD_DES;
				break;
			    case '2':
				ciph=CFS_THREE_DES;
				break;
			    case '3':
				ciph=CFS_TRUE_THREE_DES;
				break;
			    case 'b':
				ciph=CFS_BLOWFISH;
				fprintf(stderr,
				      "Warning: Blowfish is a new cipher.\n");
				break;
			    case 'm':
				ciph=CFS_MACGUFFIN;
				fprintf(stderr,
				      "Warning: MacGuffin is a weak cipher\n");
				break;
			    case 's':
				ciph=CFS_SAFER_SK128;
				fprintf(stderr,
				      "Warning: SAFER is a new cipher.\n");
				break;
			    case 'o':
				cfmt=0;
				break;
			    case 'p':	/* puny memory option */
				smsize=SMALLSMSIZE;
				break;
			    case '-':
				keycheck=0;
				break;
			    default:
				fprintf(stderr,"usage: cmkdir [-123bdmosp-] dir\n");
				exit(1);
			}
	}
	if (argc!=1) {
		fprintf(stderr,"Usage: cmkdir [-123bdmosp-] dir\n");
		exit(1);
	}
	if (keycheck) {
		if ((pw=getpassword("Key:"))==NULL) {
			fprintf(stderr,"Can't get key\n");
			exit(1);
		}
		strcpy(pword,pw);
		if (strlen(pw)<16) {
			fprintf(stderr,"Key must be at least 16 chars.\n");
			exit(1);
		}
		if ((pw=getpassword("Again:"))==NULL) {
			fprintf(stderr,"Can't get key\n");
			exit(1);
		}
		if (strcmp(pword,pw)!=0) {
		     fprintf(stderr,
			"Keys don't match; drink some coffee and try again\n");
		     exit(1);
		}
	}
	else { /* just accept key from stdio */
		if (fgets(pword,256,stdin) == NULL) {
			perror("cmkdir");
			exit(1);
		}
		pw=pword;
		pw[255]='\0';
		n=strlen(pw);
		if ((n>0) && (pw[n-1] == '\n'))
			pw[n-1] = '\0';
	}
	if (smsize != LARGESMSIZE)
		sprintf(pw,"%s%d",pword,smsize);
	k.cipher=ciph;
	if (cfmt==0) { 
		if (old_pwcrunch(pw,&k)!=0) {
			fprintf(stderr,"Invalid key\n");
			exit(1);
		}
	} else {
		/* this is very ugly and will be replaced but it works */
		if (new_pwcrunch(pw,&k)!=0) {
			fprintf(stderr,"Invalid key\n");
			exit(1);
		}
		/* now we xor in some truerand bytes for good measure */
		bcopy(&k,ekey,32);  /* assumes key material < 32 bytes */
		for (i=0; i<32; i++) {
			ekey[i] ^= randbyte(); 
		}
		encrypt_key(&k,ekey);
		bcopy(ekey,ek1,32);
		decrypt_key(&k,ek1);
		/* new &k is our real key */
	}
	if (mkdir(argv[0],0777)<0) {
		perror("cmkdir");
		exit(1);
	}
	sprintf(path,"%s/...",argv[0]);
	strcpy(str,"qua!");
	/* now randomize the end of str.. */
	r = trand32();
	for (i=0; i<4; i++)
		str[i+4]=(r<<(i*8))&0377;
	copykey(&k,&kt);
	cipher(&kt,str,0);
	mask_cipher(&kt,str,1);
	cipher(&kt,str,0);
	if ((fp=fopen(path,"w")) == NULL) {
		perror("cmkdir");
		exit(1);
	}
	fwrite(str,8,1,fp);
	fclose(fp);
	sprintf(path,"%s/..c",argv[0]);
	if ((fp=fopen(path,"w")) == NULL) {
		perror("cmkdir");
		exit(1);
	}
	fprintf(fp,"%d",k.cipher);
	fclose(fp);
	sprintf(path,"%s/..s",argv[0]);
	if ((fp=fopen(path,"w")) == NULL) {
		perror("cmkdir");
		exit(1);
	}
	fprintf(fp,"%d\n",smsize);
	fclose(fp);
	if (cfmt) {
		sprintf(path,"%s/..k",argv[0]);
		if ((fp=fopen(path,"w")) == NULL) {
			perror("cmkdir");
			exit(1);
		}
		fwrite(ekey,32,1,fp);
		fclose(fp);
	}
	exit(0);
}
